src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1983,10 +2077,32 @@ instance Semigroup LimitClause where
 
 instance Monoid LimitClause where
   mempty = Limit mzero mzero
-  mappend = (<>)
 
 -- | A locking clause.
-type LockingClause = Monoid.Last LockingKind
+data LockingClause =
+  GeneralLockingClause LockingKind
+  -- ^ Locking clause not specific to any database implementation
+  | PostgresLockingClauses [PostgresLockingKind]
+  -- ^ Locking clause specific to postgres
+  | NoLockingClause
+
+instance Semigroup LockingClause where
+  -- Postgres allows us to have multiple locking clauses
+  (<>) (PostgresLockingClauses pleft) (PostgresLockingClauses pright) = PostgresLockingClauses (pleft <> pright)
+  -- For backwards compatibility reasons we use the monoid instance of last for
+  -- general locking cluase
+  (<>) (GeneralLockingClause gleft) (GeneralLockingClause gright)  =
+    Maybe.maybe mempty GeneralLockingClause
+      (Monoid.getLast $ Monoid.Last (Just gleft) <> Monoid.Last (Just gright))
+
+  (<>) gleft@(GeneralLockingClause _) (PostgresLockingClauses _)  = gleft
+  (<>) (PostgresLockingClauses _) gright@(GeneralLockingClause _) = gright
```

Comment:
I feel pretty unsatisfied here. I can imagine that this would be *very* confusing to debug if it comes up in a production context. Imagine looking at a `SqlQuery` that has locking behavior you want, and composing that with another `SqlQuery`. Later on, someone adds a `GeneralLockingClause` to the other query, and suddenly you lose the locking properties in the query you care about. The `GeneralLockingClause` destroys the `PostgresLockingClause` information, and the only information you'd get is that the query silently doesn't do what you want.

🤔 

It's especially challenging because the `GeneralLockingClause` just takes the `Last` behavior - so if you had `foo >> bar`, where `foo` does a `GeneralLockingClause`, then `bar`'s own locking clause would *overwrite* `foo` if it is a `GeneralLockingClause`. But if `bar` is a `PostgresLockingClause`, then it does the opposite thing - `foo` wins and `bar` loses.

Can we always favor the last locking clause - ie,

```haskell
    PostgresLockingClause l <> PostgresLockingClause r = 
        PostgresLockingClause (l <> r)
    l <> NoLockingClause =
        l
    l <> r = 
        r
```

I'm not sure that satisfies the monoid/semigroup laws.

I wonder if it is reasonable to collect a `GeneralLockingClause` in with the `PostgresLockingClause`? I wish I were more familiar with the interaction of these features.

Reply:
I've shifted to taking the `Last` when choosing between general and postgres locking clauses (aggregating them together doesnt work well since some of the "general" ones are rdbms specific and consequently unsafe- I've renamed it `LegacyLockingClause` to better reflect that and I would probably argue in a future release we only allow one type of locking clause)